### DISCUSSION TITLE
Using an external text file when working with Document Builder

### DISCUSSION THREAD

**Discussion by malikov-pro** [ONLYOFFICE/DocumentBuilder#60] [Q&A]
How can I write the result of the work to a text file and read the parameters from a text file when using Document Builder?

I want to get the structure of the document and pass it to the control program. When forming a document, I want to use a file with parameters, it turns out to be cumbersome to write to the command line.

**Comment by ShockwaveNN** [comment]
Could you explain that you do in more details?

>  result of the work to a text file 

Usually result of work is docx\xlsx\pptx\pdf file

Do you want to save those files as txt? Without formagging at all?

> read the parameters from a text file

What parameters for example

**Reply by malikov-pro** [reply]
> Could you explain that you do in more details?

I want to make an analog of working with the google doc service, job description https://infostart.ru/1c/articles/1283766/, repo https://github.com/malikov-pro/google_api_1c.

The first function I want to get the JSON structure of the document to check the correctness of the filling on the side of the control system.

```
{
    "status": 200,
    "data": [
        {
            "index": 0,
            "match": [
                "{v8 Область.Шапка}"
            ],
            "type": "PARAGRAPH"
        },
        ...
        {
            "index": 4,
            "match": null,
            "type": "PARAGRAPH"
        },
        {
            "index": 5,
            "match": [
                "{v8 НомерЗаказаКлиента}",
                "{v8 ДатаЗаказаКлиента}"
            ],
            "type": "PARAGRAPH"
        },
        ...
        {
            "index": 17,
            "match": [
                "{v8 ЦенаПозицииЗаказа}"
            ],
            "type": "TABLE"
        },
        ...
    ]
}
```

The second request is sending JSON with formatting parameters
```
{
  "templateDocId": "{{template_doc_id}}",
  "type": "mergeDocFromTemplate",
  "folderId": "",
  "mergeParameters": [
    {
      "index": 4,
      "replacements": []
    },
    {
      "index": 5,
      "replacements": [
        {
          "type": "text",
          "searchPattern": "{v8 НомерЗаказаКлиента}",
          "text": "23"
        }
      ]
    },
    {
      "index": 17,
      "replacements": [
        {
          "type": "image",
          "searchPattern": "{v8 ЦенаПозицииЗаказа}",
          "image": "{{image_base64}}",
          "imageType": "image/png",
          "height": 50,
          "width": 50
        }
      ]
    }
  ]
}
```

The file is needed as a transfer point between .docbuilder and node.js for example

**Reply by ShockwaveNN** [reply]
> JSON structure of the document to check the correctness of the filling on the side of the control system.

I don't thinkg there is no such method in our codebase - I think you can just parse docx file via any available parser and extract it to json only data you need, because in your example a lot of data from ooxml structure is hidden, it's a major simplification

**Reply by malikov-pro** [reply]
> I don't thinkg there is no such method in our codebase - I think you can just parse docx file via any available parser and extract it to json only data you need, because in your example a lot of data from ooxml structure is hidden, it's a major simplification

I just want to remove the docx analyzer from the control system, in my case it is 1Сv8.
The data structure contains the blocks from which the document is assembled, this was enough for product use.

**Reply by ShockwaveNN** [reply]
> docx analyzer from the control system

I don't understand why you're using docx analyzer in first place, docx structure is rather simple, it's a bunch of xml files

**Reply by malikov-pro** [reply]
> I don't understand why you're using docx analyzer in first place, docx structure is rather simple, it's a bunch of xml files

Because I will collect the document with builder objects, and not the result of parsing docx with an external component. What I know inside xml, an example of use https://infostart.ru/public/1171429/. The main question is: how to write/read data from the .docbuilder sandbox to an external system? I considered code generation as an alternative .docbuilder, but it is relatively difficult for me.

**Reply by ShockwaveNN** [reply]
I don't have any exprience in 1c, even I russian speaker, for me all [this post ](https://infostart.ru/public/1171429/) is a bucnh of unkonwn words and even unknown concepts

>  data from the .docbuilder sandbox to an external system

Put it in other way - the only way to work with DocumentBuilder is to build .js file, there is no other way 

You can try to dig inside via constuction like this:

```
builder.CreateFile("docx");
var oDocument = Api.GetDocument();
console.log(oDocument.GetClassType());
```

But only documented methods https://api.onlyoffice.com/docbuilder/textdocumentapi/apidocument/getclasstype like this is allowed

**Reply by ShockwaveNN** [reply]
I see one of possible solution is ability to generate DocBuilder js file from existing documents and editing them inline, I've created issue 51493 in our private issue tracker, currently it's not possible
